src/analysis/inner_product_space/lax_milgram.lean

```diff
@@ -0,0 +1,157 @@
+/-
+Copyright (c) 2022 Daniel Roca González. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Daniel Roca González.
+-/
+import analysis.inner_product_space.projection
+import analysis.inner_product_space.dual
+import analysis.normed_space.banach
+
+/-!
+# The The Lax-Milgram Theorem
+
+
+We consider an Hilbert space `V` over `ℝ`
+equipped with a bilinear form `B : V →L[ℝ] V →L[ℝ] ℝ`.
+For every such form we define the
+`lax_milgram_map : V →L[ℝ] V`,
+which is defined by the property `B (lax_milgram_map B v) w = ⟪v, w⟫`
+using the Fréchet-Riesz representation theorem.
+
+Under the hypothesis that `B` is *coercive* (fulfills `is_coercive`),
+we prove the Lax-Milgram theorem:
+that is, the `lax_milgram_map` can be upgraded to a continuous equivalence `lax_milgram`
+
+## References
+
+TODO cite the slides
+
+
+## Tags
+
+dual, Lax-Milgram
+-/
+
+noncomputable theory
+
+
+
+open_locale classical complex_conjugate
+open is_R_or_C continuous_linear_map linear_map linear_equiv
+
+universe u
+def is_coercive {V : Type u} [inner_product_space ℝ V] [complete_space V] (B : V →L[ℝ] V →L[ℝ] ℝ) : Prop :=
+∃ C, (0 < C) ∧ ∀ u, C * ∥u∥ * ∥u∥ ≤ B u u
+
+
+namespace lax_milgram
+variables {V : Type u} [inner_product_space ℝ V] [complete_space V]
+variables (B : V →L[ℝ] V →L[ℝ] ℝ)
+
+
+def lax_milgram_map (B : V →L[ℝ] V →L[ℝ] ℝ) : V →L[ℝ] V :=
+comp
+  (inner_product_space.to_dual ℝ V).symm.to_continuous_linear_equiv.to_continuous_linear_map
+  B
+
+
+lemma lax_milgram_map.apply : ∀ v w : V, B v w = inner (lax_milgram_map B v) w :=
+begin
+  intros x y,
+  dunfold lax_milgram_map,
+  simp,
+end
+
+
+lemma injective (coercive : is_coercive B) : (lax_milgram_map B).ker = ⊥ :=
+begin
+  ext v, split,
+  {
+    simp only [continuous_linear_map.mem_ker, submodule.mem_bot],
+    contrapose,
+    simp_rw [←ne.def],
+    intro v_ne_0,
+    --by_contradiction Av_eq_0,
+    simp_rw [ ←norm_pos_iff] at ⊢ v_ne_0,
+    rcases coercive with ⟨C, C_ge_0, coercivity⟩,
+    have squared : C * ∥v∥ * ∥v∥ ≤ ∥lax_milgram_map B v ∥ * ∥v∥,
+    by {
+      exact calc C * ∥v∥*∥v∥ ≤ B v v                        : coercivity v
+      ...                 = inner (lax_milgram_map B v) v : lax_milgram_map.apply B v v
+      ...                 ≤ ∥lax_milgram_map B v ∥ * ∥v∥    : real_inner_le_norm (lax_milgram_map B v) v,
+    },
+    have coerced := (mul_le_mul_right v_ne_0).mp squared,
+    exact calc 0 < C * ∥v∥                  : mul_pos C_ge_0 v_ne_0
+    ...          ≤ ∥(lax_milgram_map B) v∥  : coerced,
+  },
+  {
+    simp only [linear_map.mem_ker, submodule.mem_bot],
+    intro h, rw h,
+    exact linear_map.map_zero (lax_milgram_map B),
+  }
+end
+
+
+lemma closed_range : @is_closed V infer_instance ↑(lax_milgram_map B).range :=
```

Comment:
```suggestion
lemma closed_range : is_closed ((lax_milgram_map B).range : set V) :=
```
Coercion with `↑` typically only works if there's something around as a hint to what type it should be coerced to.